### ISSUE TITLE
Asciidoc-py3 project aims relative to AsciiDoctor?

### ISSUE THREAD

**Issue by maphew** [asciidoc/asciidoc-py3#67]
It would be helpful for people new to AD to have a statement or paragraph in the Readme or elsewhere that states what asciidoc-py3 is relative to asciidoctor. I'm not looking for a contest, just information on where ad-py3 might be more appropriate and what the project goals are. For example adopting one or more of the highlighted [differences](https://asciidoctor.org/docs/asciidoc-asciidoctor-diffs/)? I'm personally aligned with python, but the Dr looks more polished/solid. Add some information to help me choose the best tool for the job. Thanks!

**Comment by elextr** [comment]
Asciidoc-py3 is a port of the _original_ Asciidoc Python 2 implementation to Python 3 as it says on the repository title.  

As such it isn't "relative to asciidoctor", its intended to be the same as the original Asciidoc.  So its usable by those with old documents or by those with significant investments in tooling specific to Asciidoc that won't port to Asciidoctor (one being Ruby and one being Python, and the configuration formats being different).

It not a competitor to Asciidoctor and isn't likely to try to adopt changes due to the backward compatibility target.  

So if you are starting a new project its almost certainly better to start with Asciidoctor, its heavily supported and usually faster, and the implementation language shouldn't matter unless you are going to hack extensions and tooling that will tie you to that specific implementation whichever you choose.

**Comment by maphew** [comment]
Thank you @elextr , that's exactly what I needed.

**Comment by hoadlck** [comment]
From @elextr's comment earlier...
> It not a competitor to Asciidoctor and isn't likely to try to adopt changes due to the backward compatibility target.
> So if you are starting a new project its almost certainly better to start with Asciidoctor, its heavily supported and usually faster, and the implementation language shouldn't matter unless you are going to hack extensions and tooling that will tie you to that specific implementation whichever you choose.

I think that there are cases where where having some of the extensions from the AsciiDoctor development available in the Python implementation would be useful.  The goodness of the development to make Python AsciiDoc work like other Python modules is exactly so it can be integrated into other **Python** projects.

Likewise, it is possible to add new features and still maintain backwards compatibility.  

While it is true that AsciiDoctor is the superior implementation right now, is there support/appetite to make changes to Python AsciiDoc to achieve feature parity?  Would PRs to add new features (e.g. TOC generation without Javascript) be accepted?

I appreciate that open source projects are fueled by the free time of people that have lives and jobs.  Even if other people are putting work into PRs, there may still be a significant effort in keeping focus on the direction of the project.  Is there desire to move forward?  Or is the goal to maintain what exists?

**Comment by elextr** [comment]
> Likewise, it is possible to add new features and still maintain backwards compatibility.

Yes, some features are pure additions and don't conflict with the existing features, they would likely be fine.  It might be good for them to be off when `asciidoc` is used normally so olde documents don't suddenly start recognising something as markup that they didn't before.  But thats a normal configuration management issue that should be addressed when extra markup is recognised.

> While it is true that AsciiDoctor is the superior implementation right now, is there support/appetite to make changes to Python AsciiDoc to achieve feature parity? Would PRs to add new features (e.g. TOC generation without Javascript) be accepted?

That may be more difficult than immediately apparent, see [here](https://github.com/asciidoc/asciidoc-py3/issues/84#issuecomment-548163118) about being a streaming converter.

> I appreciate that open source projects are fueled by the free time of people that have lives and jobs. Even if other people are putting work into PRs, there may still be a significant effort in keeping focus on the direction of the project. Is there desire to move forward? Or is the goal to maintain what exists?

See the start of the issue linked above (at the end it narrowed to a technical detail, but the start addressed some of these).  @MasterOdin probably one for you to answer.

**Comment by MasterOdin** [comment]
While I would like to move towards making asciidoc more feature-parity on asciidoctor, and experiment with things like conversion of a streaming parser into a non-streaming parser (as @elextr has stated, the computing requirement of 2020 are vastly different than 2002), my immediate focus is purely technical-debt things I view in the project.

My internal milestones for the project is:
1. Move off a-a-p as a build system to strictly make
1. Fix asciidoc's "broken" method of self-recursive calls through subprocess on itself (as this makes putting it into a proper python module janky).
1. Make asciidoc.py not so heavily use globals. This goes hand-in-hand with the above, but it also makes things possible to actually test not just in an integration fashion.
1. As part of the above, start breaking apart asciidoc.py (and a2x.py to some extent) into separate files that are imported and used. This would allow for an easier focus on unit testing of asciidoc features and to raise my own confidence on changes made, or how certain things work.
1. Move asciidoc to a proper python setup.py module. While in theory the `repackage` branch works, and https://test.pypi.org/project/asciidoc/ can be installed via pip, it under the hood is not a great story (reliance on __main__.py file, which in-turn affects ability to do any sort of module imports, and the weird globals and self-recursive subprocess usage) that I plan to keep that as a reference, but probably not release it as-is.
